### PR TITLE
[fix](storage) Fix query result error due to find code by bound

### LINKED ISSUE #8315: [Enhancement] Low cardinality string optimization in storage layer
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

For the current low-cardinality string column, when loading data in the storage layer, it is necessary to traverse all the codes in the data_page, convert it into the original string value through the dictionary in the dict_page, and then insert it into the ColumnString column. Subsequent predicate calculation, filtering, etc. are based on ColumnString for processing.

If you continue to use encoding during predicate calculation, it is equivalent to converting string comparisons to int type comparisons, resulting in greater performance and less memory consumption.

### Solution

Storage layer read call: `SegmentIterator::next_batch-> ... -> PageDecoder::next_batch`, the default batch reads 4096 rows, PageDecoder reads part of data_page, data_page may be bitshuffle encoding or plain encoding.

After enabling low cardinality optimization, optimize for string columns with predicate calculations. The resulting column is of type ColumnDictionary and is used to store encodings and dictionaries. If the current batch of data_pages are all bitshuffle encoded, that is, directly insert the encoded value and dictionary into the result column, in the process of PageDecode next_batch, the encoding to String overhead (there is the overhead of constructing a dictionary) is saved. After the column data is loaded in the Segment layer, the predicate calculation is performed based on the encoded value and the dictionary, that is, the predicate constant is converted into an int value through the dictionary, and the int value is compared with the int in the encoded column, which is equivalent to comparing the original String. Converted to int comparison, there will be a significant speedup effect. After the predicate is calculated, the encoding column and dictionary are converted into ColumnString columns to end the segment layer reading.

If there is plain encoded data_page in the current batch, the dictionary cannot be used and is converted to a PredicateColumn column for processing.

### Performance
env: 1 FE, 3 BE
version: c66a9bf

| Id        | Doris vectorization (including storage layer)                | After optimization                                           | Remark |
| --------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ---- |
| Q2.1 flat | - Total: 769ms<br/>    - BlockLoadTime: 10s636ms<br/>    - BlockSeekTime: 3s199ms<br/>        - ReadColByIdxTime: 3s732ms<br/>        - **ShortPredEvalTime: 3s31ms**<br/>        - OutputColBySelIdxTime: 72.968ms<br/>        - OutputNonPredColTime: 9.814ms<br/>        - ReadColByRowIdTime: 3s589ms<br/>        - InitCurBlockTime: 42.203ms<br/>        - VectorPredEvalTime: 104.169ms | - Total: 427ms<br/>    - BlockLoadTime: 5s720ms<br/>    - BlockSeekTime: 1s798ms<br/>        - ReadColByIdxTime: 3s32ms<br/>        - **ShortPredEvalTime: 277.569ms**<br/>        - OutputColBySelIdxTime: 67.17ms<br/>        - OutputNonPredColTime: 6.655ms<br/>        - ReadColByRowIdTime: 2s171ms<br/>        - InitCurBlockTime: 37.303ms<br/>        - VectorPredEvalTime: 98.790ms |      |
| Q2.2 flat | - Total: 645ms<br/>    - BlockLoadTime: 9s352ms<br/>    - BlockSeekTime: 1s145ms<br/>        - ReadColByIdxTime: 2s957ms<br/>        - **ShortPredEvalTime: 5s10ms**<br/>        - OutputColBySelIdxTime: 27.5ms<br/>        - OutputNonPredColTime: 7.275ms<br/>        - ReadColByRowIdTime: 1s193ms<br/>        - InitCurBlockTime: 32.54ms<br/>        - VectorPredEvalTime: 92.931ms | - Total: 407ms<br/>    - BlockLoadTime: 5s385ms<br/>    - BlockSeekTime: 1s365ms<br/>        - ReadColByIdxTime: 3s219ms<br/>        - **ShortPredEvalTime: 559.475ms**<br/>        - OutputColBySelIdxTime: 31.517ms<br/>        - OutputNonPredColTime: 7.107ms<br/>        - ReadColByRowIdTime: 1s408ms<br/>        - InitCurBlockTime: 28.264ms<br/>        - VectorPredEvalTime: 94.499ms |      |
| Q2.3 flat | - Total: 511ms<br/>    - BlockLoadTime: 6s997ms<br/>    - BlockSeekTime: 939.491ms<br/>        - ReadColByIdxTime: 2s831ms<br/>        - **ShortPredEvalTime: 3s65ms**<br/>        - OutputColBySelIdxTime: 12.812ms<br/>        - OutputNonPredColTime: 7.697ms<br/>        - ReadColByRowIdTime: 937.718ms<br/>        - InitCurBlockTime: 19.722ms<br/>        - VectorPredEvalTime: 93.701ms | - Total: 352ms<br/>    - BlockLoadTime: 4s748ms<br/>    - BlockSeekTime: 1s218ms<br/>        - ReadColByIdxTime: 3s105ms<br/>        - **ShortPredEvalTime: 258.50ms**<br/>        - OutputColBySelIdxTime: 15.446ms<br/>        - OutputNonPredColTime: 6.700ms<br/>        - ReadColByRowIdTime: 1s211ms<br/>        - InitCurBlockTime: 20.622ms<br/>        - VectorPredEvalTime: 95.373ms |      |

In the SSB fact-table test Q2.x, after the low-cardinality string optimization is completed, the short-circuit predicate calculation performance is improved by about 90%, the storage layer is improved by about 40%, and the overall improvement is about 37%.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zenoyang**
# Proposed changes

Issue Number: close #8315 
Fix query result error due to find code by bound.

## Problem Summary:

Problem recurrence
SSB single table `lineorder_flat`, the query SQL is as follows:
```sql
SELECT
    sum(LO_REVENUE),
    (LO_ORDERDATE DIV 10000) AS year,
    P_BRAND
FROM lineorder_flat
WHERE P_BRAND >= 'MFGR#22211111' AND P_BRAND <= 'MFGR#22281111' AND S_REGION = 'ASIA' and (LO_ORDERDATE DIV 10000) = 1992
GROUP BY
    year,
    P_BRAND
ORDER BY
    year,
    P_BRAND;
```

when `enable_low_cardinality_optimize=false`, query result：
```sql
+-------------------+------+-----------+
| sum(`LO_REVENUE`) | year | P_BRAND   |
+-------------------+------+-----------+
|       65423264312 | 1992 | MFGR#2222 |
|       66936772687 | 1992 | MFGR#2223 |
|       64047191934 | 1992 | MFGR#2224 |
|       65744559138 | 1992 | MFGR#2225 |
|       66993045668 | 1992 | MFGR#2226 |
|       67411226147 | 1992 | MFGR#2227 |
|       69390885970 | 1992 | MFGR#2228 |
+-------------------+------+-----------+
```

when `enable_low_cardinality_optimize=true`, query result：
```sql
+-------------------+------+-----------+
| sum(`LO_REVENUE`) | year | P_BRAND   |
+-------------------+------+-----------+
|       66936772687 | 1992 | MFGR#2223 |
|       64047191934 | 1992 | MFGR#2224 |
|       65744559138 | 1992 | MFGR#2225 |
|       66993045668 | 1992 | MFGR#2226 |
|       67411226147 | 1992 | MFGR#2227 |
|       69390885970 | 1992 | MFGR#2228 |
+-------------------+------+-----------+
```

One line less than the correct result.

The reason is that 'MFGR#22211111' is not in the dictionary, so get the boundary code (`find_code_by_bound` method), but there is a bug here.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/in_list_predicate.cpp [hunk 1]
```diff
@@ -299,8 +299,7 @@ IN_LIST_PRED_BITMAP_EVALUATE(NotInListPredicate, -=)
                 auto& dict_col =                                                               \
                         reinterpret_cast<vectorized::ColumnDictionary<vectorized::Int32>&>(    \
                                 *col_ptr);                                                     \
-                auto code_set = dict_col.find_codes(_values);                                  \
-                _dict_codes = std::move(code_set);                                             \
+                _dict_codes = std::move(dict_col.find_codes(_values));                         \
```
Comment (HappenLee) [inline be/src/olap/in_list_predicate.cpp:302]: not need use `std::move`，it's right value

File: be/src/olap/comparison_predicate.cpp [hunk 2]
```diff
@@ -548,8 +547,7 @@ COMPARISON_PRED_SET_DICT_CODE(NotEqualPredicate)
                 auto& dict_col =                                                               \
                         reinterpret_cast<vectorized::ColumnDictionary<vectorized::Int32>&>(    \
                                 *col_ptr);                                                     \
-                auto code = dict_col.find_code_by_bound(_value, 0 OP 1, 1 OP 1);               \
-                _dict_code = code;                                                             \
+                _dict_code = dict_col.find_code_by_bound(_value, 1 OP 0, 1 OP 1);              \
```
Comment (HappenLee) [inline be/src/olap/comparison_predicate.cpp:550]: what means of 1,0 ? maybe we need more comment?

Comment (zenoyang) [reply be/src/olap/comparison_predicate.cpp:550]: If `1 OP 0` returns true, it means the predicate is `>` or `>=`,
If `1 OP 1` returns true, it means the predicate is `>=` or `<=`
Through this trick, avoid redundant code, such as `bool is_greater = type() == PredicateType::GT || type() == PredicateType::GE;`
ok, i will add a comment.

File: be/src/olap/in_list_predicate.cpp (hunk 1 above)
Comment (zenoyang) [reply be/src/olap/in_list_predicate.cpp]: done

File: be/src/vec/columns/column_dictionary.h [hunk 3]
```diff
@@ -294,19 +294,14 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
...
             }
-
-            if (lower) {
-                return std::lower_bound(_dict_data.begin(), _dict_data.end(), value) -
-                       _dict_data.begin() - eq;
-            } else {
-                return std::upper_bound(_dict_data.begin(), _dict_data.end(), value) -
-                       _dict_data.begin() + eq;
-            }
+            auto bound = std::upper_bound(_dict_data.begin(), _dict_data.end(), value) -
```
Comment (wangbo) [inline be/src/vec/columns/column_dictionary.h:302]: Need some comment to make here easier to understand

Comment (zenoyang) [reply be/src/vec/columns/column_dictionary.h:319]: done

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**